csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs

```diff
@@ -173,23 +167,96 @@ public object GetProperty(string propertyName)
 
             if (propertyType == PropertyType.Int)
             {
-                var longPropertyValue = Marshal.ReadInt64(propertyValue);
-                allocator.FreeMemory(propertyValue);
-                return longPropertyValue;
+                Int64 value;
+                unsafe
+                {
+                    value = *(Int64*)propertyValue;
+                }
+                return value;
             }
             else if (propertyType == PropertyType.Float)
             {
-                float[] value = new float[1];
-                Marshal.Copy(propertyValue, value, 0, 1);
-                allocator.FreeMemory(propertyValue);
-                return value[0];
+                float value;
+                unsafe
+                {
+                    value = *(float*)propertyValue;
+                }
+                return value;
             }
             else if (propertyType == PropertyType.String)
             {
                 return NativeOnnxValueHelper.StringFromNativeUtf8(propertyValue, allocator);
             }
 
-            throw new ArgumentException("Expected the property type to be one of long, float or string. Unknown type retrieved " + propertyValue.ToString());
+            try
+            {
+                throw new ArgumentException("Expected the property type to be one of long, float or string. Unknown type retrieved " + propertyValue.ToString());
+            }
+            finally
+            {
+                allocator.FreeMemory(propertyValue);
+            }
```

Comment:
This try block is almost never entered in, and finally therefore almost never runs.

Reply:
It is expected that the code does not reach this point since it would imply that we have not added support for new types in the C# binding but have the support in C native function.
Was there a recommendation here on how to deal with this in a better way?

Reply:
The issue here, is that proeprtyValue must always be deallocated before returning from the function. But it is not deallocated for int and float datatypes since FreeMemory is never reached. I suggest to not to pass allocator to `StringFromNativeUtf8`. Instead, wrap the whole block into try finally with `FreeMemory in it`. The `ArgumentException` can be thrown at the end of the `try` block.

Reply:
Addressed this. Thanks for the suggestion.